src/papilo/io/PboParser.hpp

```diff
@@ -0,0 +1,397 @@
+          std::move( Vec<ColFlags> vect( n, kIntegral ) ) );
+      problem.setVariableNames( std::move( parser.colnames ) );
+      problem.setName( std::move( filename ) );
+      // We do not have ConstraintNames in PBO
+      // problem.setConstraintNames( std::move( parser.rownames ) );
+
+      /* Not sure what to do with InputTolerance
+         problem.setInputTolerance(
+            REAL{ pow( typename RealParseType<REAL>::type{ 10 },
+                       -std::numeric_limits<
+                           typename RealParseType<REAL>::type>::digits10 ) } );
+       */
+      return problem;
+   }
+
+ private:
+   PboParser() {}
+
+   /// load LP from PBO file as transposed triplet matrix
+   bool
+   parseFile( const std::string& filename );
+
+   bool
+   parse( boost::iostreams::filtering_istream& file );
+
+   /* Try to comply with http://www.cril.univ-artois.fr/PB16/format.pdf
+    * but not rely on competition specific rules
+    * data for pbo problem
+    */
+
+   Vec<Triplet<REAL>> entries;
+   Vec<std::pair<int, REAL>> coeffobj;
+   Vec<REAL> rowlhs;
+   Vec<REAL> rowrhs;
+   Vec<std::string> colnames;
+
+   HashMap<std::string, int> colname2idx;
+   Vec<RowFlags> row_flags;
+   REAL objoffset = REAL( 0 );
+
+   int nCols = 0;
+   int nRows = 0;
+   int nnz = -1;
+
+   Vec<std::pair<int, REAL>>
+   parseRow( std::string& trimmedstrline );
+};
+
+template <typename REAL>
+bool
+PboParser<REAL>::parseFile( const std::string& filename )
+{
+   std::ifstream file( filename, std::ifstream::in );
+   boost::iostreams::filtering_istream in;
+
+   if( !file )
+      return false;
+
+#ifdef PAPILO_USE_BOOST_IOSTREAMS_WITH_ZLIB
+   if( boost::algorithm::ends_with( filename, ".gz" ) )
+      in.push( boost::iostreams::gzip_decompressor() );
+#endif
+
+#ifdef PAPILO_USE_BOOST_IOSTREAMS_WITH_BZIP2
+   if( boost::algorithm::ends_with( filename, ".bz2" ) )
+      in.push( boost::iostreams::bzip2_decompressor() );
+#endif
+
+   in.push( file );
+
+   return parse( in );
+}
+
+template <typename REAL>
+std::pair<Vec<std::pair<int, REAL>>, REAL>
+parseRow( std::string& trimmedstrline )
+{
+   const std::string whitespace = " ";
+   auto beginSpace = trimmedstrline.find_first_of( whitespace );
+   while( beginSpace != std::string::npos )
+   {
+      const auto endSpace =
+          trimmedstrline.find_first_not_of( whitespace, beginSpace );
+      const auto range = endSpace - beginSpace;
+
+      trimmedstrline.replace( beginSpace, range, fill );
+
+      const auto newStart = beginSpace + fill.length();
+      beginSpace = trimmedstrline.find_first_of( whitespace, newStart );
+   } // having a nicer string to work with makes me comfortable i get it right.
+     // Loop maybe O(n^2)
+   Vec<std::pair<int, REAL>> result;
+
+   REAL rhsoff = REAL( 0 );
+
+   std::stringstream row( beginSpace );
+
+   int degree = 0;
+   int variable_index;
+   REAL weight;
+
+   // I am unfamiliar with error handling conventions in this code base.
```

Comment:
If an error or unexpected data structure is occurring I think is best to abort to signal the user that the code does not support the code 100%